text/0975-add-context-api.md

```diff
@@ -0,0 +1,148 @@
+---
+stage: accepted
+start-date: 2023-09-28T00:00:00.000Z # In format YYYY-MM-DDT00:00:00.000Z
+release-date: # In format YYYY-MM-DDT00:00:00.000Z
+release-versions:
+teams: # delete teams that aren't relevant
+  - framework
+prs:
+  accepted: https://github.com/emberjs/rfcs/pull/975
+project-link:
+suite:
+---
+
+<!---
+Directions for above:
+
+stage: Leave as is
+start-date: Fill in with today's date, 2032-12-01T00:00:00.000Z
+release-date: Leave as is
+release-versions: Leave as is
+teams: Include only the [team(s)](README.md#relevant-teams) for which this RFC applies
+prs:
+  accepted: Fill this in with the URL for the Proposal RFC PR
+project-link: Leave as is
+suite: Leave as is
+-->
+
+# Add context API
+
+## Summary
+Add a context API to allow sharing state with all descendants of a component
+without prop-drilling.
+
+## Motivation
+Ember provides developers with great APIs for sharing state and props. Services
+allow us to set global, application-wide state. Contextual components allow us
+to expose components pre-filled with data.
+
+However, the available solutions aren't enough for situations where we want to
+access local state in deeply nested component trees. Services are global, but
+what if we want to render two component trees side by side, and have each of
+them expose a different "context state" to all descendants?
+
+Similarly, while yielding does give us the ability to pass components or state
+around, we'd still have to explicitly add arguments to all components rendered
+within a section of the app.
+
+Alternatively, we can also "prop-drill" - pass certain arguments through to all
+components, sometimes just so that one deeply nested component can access the
+information.
+
+This is where a context API would greatly improve the developer experience. Some
+examples of where context might be better than existing solutions.
+
+- A design system
+
+  In a component library, a `Card` component might accept a `backgroundColor`
+argument. The same component could then also expose this background color in its
+context state, making it available to all components nested inside it - no
+matter how deep.
+
+  Other library components, like buttons or text components, could then check what
+background color they're rendered on, and adjust their own styles accordingly,
+to make sure the text and background color contrasts are accessible, or just to
+make sure the button background matches the card background ("danger"-style
+buttons in "danger"-styled cards). These components could adjust their color no
+matter how deeply nested they are inside the card, without us having to
+explicitly pass the background color through multiple layers of components.
+
+- Chart and graphical components
```

Comment:
Another use case is _drastically_ simplifying and improving ember-kepboard, for having real hierarchical shortcuts